flink-runtime/src/test/java/org/apache/flink/runtime/source/coordinator/SourceCoordinatorAlignmentTest.java

```diff
@@ -131,6 +138,101 @@ void testWatermarkAlignmentWithTwoGroups() throws Exception {
         }
     }
 
+    @Test
+    @Timeout(5)
+    void testWatermarkAggregatorBenchmark() {
```

Comment:
Added a unit test about benchmark:

- The duration of this test based on master branch is 1 min 44s
- The duration of this test based on master branch with this PR is 469ms

The performance is improved by almost 200 times.

<img width="905" alt="image" src="https://github.com/apache/flink/assets/38427477/b4346d73-50a9-4d0f-8d7e-398220f7c612">


<img width="1024" alt="image" src="https://github.com/apache/flink/assets/38427477/47ffcb55-1c57-40b9-bc12-9e3fb88e4da8">

Reply:
This benchmark test is removed in this PR, I'm not sure if we should add a benchmark in the `flink-benchmarks` repo for this case. It's a small case.

Anyway, if you think it's needed, I'm happy to do it.

Reply:
I think it would be valuable to have something like that. Otherwise we wouldn't spot a performance regression in this area.
 
Option 1. should be very easy to implement, but I dunno if we can make this optimisation actually visible in a microbenchmark on a single machine (so probably max parallelism of sth between 16 to 128)

Option 2. would be more work, but you could check how the `Scheduler` benchmarks have been implemented.  
<img width="818" alt="Screenshot 2023-06-29 at 12 53 44" src="https://github.com/apache/flink/assets/8957547/5c677291-ae2f-4eba-83fc-66be64f1e11f">
I have never looked in the code, but I wouldn't be surprised that it's a very similar construct that you would need for Option 2.

Reply:
If it's valuable, I prefer finish the benchmark first, then merge this PR. We can see the performance change from the WebUI.

About how to do the benchmark, let me think a little more~